## **4. Association Rule**

### **a)**

In [ ]:
!pip install mlxtend  

In [ ]:
!pip install --upgrade mlxtend

In [3]:
import mlxtend

### **b)**

In [4]:
import pandas as pd
from mlxtend.preprocessing import TransactionEncoder
from sklearn.preprocessing import MultiLabelBinarizer

In [5]:
transactions = pd.read_csv('Basket.csv',header=None, sep='\n',names=['col'])

In [6]:
transactions.head(3)

col
0  Lassi,Coffee Powder,Butter,Yougurt,Ghee,Cheese,
1                              Ghee,Coffee Powder,
2                  Lassi,Tea Powder,Butter,Cheese,

<hr style = "border-top: 3px solid #000000 ; border-radius: 3px;">

<p style =" direction:rtl;text-align:right;">
ابتدا فایل cvs را به کمک جداکننده \n و در یک ستون می خوانیم.
</p>

<hr style = "border-top: 3px solid #000000 ; border-radius: 3px;">


In [9]:
dataset = transactions['col'].apply(lambda dataset: dataset.split(','))
dataset = list(dataset)
te = MultiLabelBinarizer()
te_ary = te.fit(dataset).transform(dataset)
df = pd.DataFrame(te_ary, columns=te.classes_)
df.head(3)

Bread  Butter  Cheese  Coffee Powder  Ghee  Lassi  Milk  Panner  Sugar  \
0  1      0       1       1              1     1      1     0       0      0   
1  1      0       0       0              1     1      0     0       0      0   
2  1      0       1       1              0     0      1     0       0      0   

   Sweet  Tea Powder  Yougurt  
0      0           0        1  
1      0           0        0  
2      0           1        0

<hr style = "border-top: 3px solid #000000 ; border-radius: 3px;">

<p style =" direction:rtl;text-align:right;">
 
    .سپس اعضای هر سطر را به کمک , ها جدا کرده و به کمک کلاس فوق آنها را به یک دیتاست باینری تبدیل میکنیم که 1 نشانه حضور کالا در تراکنش و 0 نشانه عدم حضور است
</p>

<hr style = "border-top: 3px solid #000000 ; border-radius: 3px;">


### **C)**

In [10]:
from mlxtend.frequent_patterns import apriori
from mlxtend.frequent_patterns import fpgrowth

In [13]:
frequent_items =apriori(df, min_support=0.1,use_colnames=True)
frequent_items

support                  itemsets
0    1.000000                        ()
1    0.437809                   (Bread)
2    0.437570                  (Butter)
3    0.437171                  (Cheese)
4    0.439805           (Coffee Powder)
..        ...                       ...
154  0.199824     (, Sweet, Tea Powder)
155  0.197829        (, Sweet, Yougurt)
156  0.198467   (, Tea Powder, Yougurt)
157  0.100990    (Sweet, Panner, Lassi)
158  0.100990  (, Sweet, Panner, Lassi)

[159 rows x 2 columns]

In [14]:
 %timeit -n 100 -r 10  apriori(df, min_support=0.1,use_colnames=True)

72.9 ms ± 4.22 ms per loop (mean ± std. dev. of 10 runs, 100 loops each)


<hr style = "border-top: 3px solid #000000 ; border-radius: 3px;">

<p style =" direction:rtl;text-align:right;">
    در خروجی فوق support  نشان دهنده فرکانس نسبی تکرار یم رابطه استو itemsets اعضای حاضر در آن رابطه را نشان می دهد.
    بدیهی است که برای اینکه مطمئن شویم یک رابطه اتفاقی پیدانشده است به دنبال support های بالاتر هستیم.
اگر use_columns را false قراردهیم اندیس ستون ها را نمایش می دهد.
</p>
<p style =" direction:rtl;text-align:right;">
 با برای مثال در نمونه فوق  (Coffee Powder) با فرکانس 0.439805 بین تراکنش ها تکرار شده اند.
    از این مجموعه ها میتوان بعدا برای کشف روابط استفاده کرد.
</p>

<hr style = "border-top: 3px solid #000000 ; border-radius: 3px;">


### **d)**

In [15]:
 %timeit -n 100 -r 10   fpgrowth(df,min_support=0.1, use_colnames=True)

119 ms ± 3.02 ms per loop (mean ± std. dev. of 10 runs, 100 loops each)


<hr style = "border-top: 3px solid #000000 ; border-radius: 3px;">

<p style =" direction:rtl;text-align:right;">
Apriori: یک الگوریتم مبتنی بر آرایه  است که از تکنیک Join و  Prune و الگوریتم  breadth-first search استفاده می کند. در این روش ابتدا الگوهای حاوی 1 مورد ، سپس 2 مورد ، سپس 3 مورد و غیره تولید می شوند و تولید کاندیداها بسیار کند است. زمان اجرا بسته به تعداد موارد مختلف به طور تصاعدی افزایش می یابد اما قابلیت موازیسازی وجود دارد اما به دلیل وجود تعداد زیادی از کاندیداها ، به فضای حافظه زیادی احتیاج دارد و دیتاست را چندین بار برای تولید مجموعه نامزدها اسکن می کند.
</p>
<p style =" direction:rtl;text-align:right;">Fpgrowth:  این الگوریتم یک  روش مبتنی بر درخت است.این درخت الگوهای تکراری را از پایگاه داده ایجاد می کند که حداقل support را برآورده می کنند و از جستجوی عمق اول استفاده می کند و  فقط الگوهای موجود در پایگاه داده را در نظر می گیرد. زمان اجرای آن بسته به تعداد تراکنش ها و موارد به صورت خطی افزایش می یابد . داده ها بسیار وابسته به هم هستند ، هر گره به ریشه نیاز دارد.به دلیل فضای فشرده و عدم  تولید همه کاندیداها ، به فضای حافظه کمتری نیاز دارد.فقط دوبار از پایگاه داده اسکن می کند تا درخت را ایجاد کند.
</p>

<hr style = "border-top: 3px solid #000000 ; border-radius: 3px;">


### **e)**

In [18]:
from mlxtend.frequent_patterns import association_rules

In [19]:
rules = association_rules(frequent_items, metric="lift", min_threshold=1)
rules.head(3)

antecedents consequents  antecedent support  consequent support   support  \
0          ()     (Bread)            1.000000            0.437809  0.437809   
1     (Bread)          ()            0.437809            1.000000  0.437809   
2          ()    (Butter)            1.000000            0.437570  0.437570   

   confidence  lift  leverage  conviction  
0    0.437809   1.0       0.0         1.0  
1    1.000000   1.0       0.0         inf  
2    0.437570   1.0       0.0         1.0

### **f)**

In [24]:
rules.iloc[29:30]

antecedents consequents  antecedent support  consequent support  \
29  (Coffee Powder)     (Bread)            0.439805            0.437809   

    support  confidence     lift  leverage  conviction  
29  0.20182    0.458885  1.04814  0.009269    1.038949

<hr style = "border-top: 3px solid #000000 ; border-radius: 3px;">

<p style =" direction:rtl;text-align:right;">
این سطر بیانگر این است که Coffee و Powder و Bread به طور مکرر خریداری شده اند .
ساپورت برای Coffee و Powder برابر 0.439805 است که فرکانس نسبی تکرار این رابطه (کاندیدا) است.
این عدد برای bread هم مجزا در قابل consequent support گزارش شده است.
support به تنهایی هردو این محموعه ها را در نظر میگیرد و برای اجتماع آنها این عدد را 0.20182 گزارش می کند.
    
    
کانفیدنس برای این قاعده 0.458885 است یعنی بین همه افرادی که Coffee و  Powder  را خریداری کرده اند 45درصد انها Bread را هم خریده اند.
    در واقع احتمال وقوع این دو پیشامد بایکدیگر است.
    
</p>

<p style =" direction:rtl;text-align:right;">


از متریک لیفت معمولاً برای اندازه گیری اینکه antecedents و در consequents یک قانون A-> C اغلب چقدر باهم اتفاق می افتد استفاده می شود ، بیش از آنکه انتظار داشته باشیم اگر از نظر آماری مستقل باشند. 
    اگر A و C مستقل باشند ، امتیاز Lift دقیقاً 1 خواهد بود.
</p>

<p style =" direction:rtl;text-align:right;">

leverage تفاوت بین فرکانس مشاهده شده A و C با هم و فرکانسی را که در صورت استقلال A و C انتظار می رود محاسبه می کند. مقدار  0 نشانگر استقلال است.

</p>

<p style =" direction:rtl;text-align:right;">

مقدار conviction زیاد به این معنی است که consequents بسیار بستگی به antecedents دارد. به عنوان مثال ، در مورد confidence کامل ، مخرج 0 می شود (به دلیل 1 - 1) که  conviction برای آن "inf" تعریف شده است. مشابه lift ، اگر اقلام مستقل باشند ، محکومیت 1 است.
</p>
<hr style = "border-top: 3px solid #000000 ; border-radius: 3px;">


<img src="conviction.png"
     alt="Markdown Monster icon"
     style="float: left; margin-right: 10px;" />
<img src="levarage.png"
     alt="Markdown Monster icon"
     style="float: left; margin-right: 10px;" />
<img src="lift.png"
     alt="Markdown Monster icon"
     style="float: left; margin-right: 10px;" />
 <img src="conf.png"
     alt="Markdown Monster icon"
     style="float: left; margin-right: 10px;" />         
 <img src="sup.png"
     alt="Markdown Monster icon"
     style="float: left; margin-right: 10px;" />         
     
